File: ALODWebUtility/LookUps/LookUpTable.vb
Imports System.Text
Imports System.Web.UI.WebControls
Imports ALOD.Data

Namespace LookUps

    Public Class LookUpTable
        Implements IList(Of LookUpRow)

        Protected _adapter As SqlDataStore
        Protected _rows As List(Of LookUpRow)
        Protected cacheSource As System.Web.Caching.Cache

        Public Sub New()
            _rows = New List(Of LookUpRow)
        End Sub

        Public Property rows() As List(Of LookUpRow)
            Get
                Return _rows
            End Get
            Set(ByVal value As List(Of LookUpRow))
                _rows = value
            End Set
        End Property

        Protected ReadOnly Property Cache() As System.Web.Caching.Cache
            Get
                Return System.Web.HttpContext.Current.Cache
            End Get
        End Property

        Protected ReadOnly Property DataStore() As SqlDataStore
            Get
                If (_adapter Is Nothing) Then
                    _adapter = New SqlDataStore()
                End If
                Return _adapter
            End Get
        End Property

        Public Sub AddRow(ByVal value As String, ByVal text As String)

            Dim row As New LookUpRow()
            row.value = value
            row.text = text
            _rows.Add(row)

        End Sub

        Public Function Find(ByVal value As String) As LookUpRow

            For Each row As LookUpRow In _rows
                If (row.value = value) Then
                    Return row
                End If
            Next

            Return Nothing

        End Function

        Public Sub LookUpRowReader(ByVal adapter As SqlDataStore, ByVal reader As IDataReader)

            Dim row As New LookUpRow()
            If (Not reader.IsDBNull(0)) Then
                row.value = reader(0).ToString()
            Else
                row.value = String.Empty
            End If

            If (Not reader.IsDBNull(1)) Then
                row.text = reader(1).ToString()
            Else
                row.text = String.Empty
            End If
            _rows.Add(row)

        End Sub

#Region "IList"

        Public ReadOnly Property Count() As Integer Implements System.Collections.Generic.ICollection(Of LookUpRow).Count
            Get
                Return _rows.Count
            End Get
        End Property

        Public ReadOnly Property IsReadOnly() As Boolean Implements System.Collections.Generic.ICollection(Of LookUpRow).IsReadOnly
            Get
                Return False
            End Get
        End Property

        Default Public Property Item(ByVal index As Integer) As LookUpRow Implements System.Collections.Generic.IList(Of LookUpRow).Item
            Get
                Return _rows(index)
            End Get
            Set(ByVal value As LookUpRow)
                _rows(index) = value
            End Set
        End Property

        Public Sub Add(ByVal item As LookUpRow) Implements System.Collections.Generic.ICollection(Of LookUpRow).Add
            _rows.Add(item)
        End Sub

        Public Sub Clear() Implements System.Collections.Generic.ICollection(Of LookUpRow).Clear
            _rows.Clear()
        End Sub

        Public Function Contains(ByVal item As LookUpRow) As Boolean Implements System.Collections.Generic.ICollection(Of LookUpRow).Contains
            Return _rows.Contains(item)
        End Function

        Public Sub CopyTo(ByVal array() As LookUpRow, ByVal arrayIndex As Integer) Implements System.Collections.Generic.ICollection(Of LookUpRow).CopyTo
            _rows.CopyTo(array, arrayIndex)
        End Sub

        Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of LookUpRow) Implements System.Collections.Generic.IEnumerable(Of LookUpRow).GetEnumerator
            Return _rows.GetEnumerator()
        End Function

        Public Function GetEnumerator1() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
            Return _rows.GetEnumerator()
        End Function

        Public Function IndexOf(ByVal item As LookUpRow) As Integer Implements System.Collections.Generic.IList(Of LookUpRow).IndexOf
            Return _rows.IndexOf(item)
        End Function

        Public Sub Insert(ByVal index As Integer, ByVal item As LookUpRow) Implements System.Collections.Generic.IList(Of LookUpRow).Insert
            _rows.Insert(index, item)
        End Sub

        Public Function Remove(ByVal item As LookUpRow) As Boolean Implements System.Collections.Generic.ICollection(Of LookUpRow).Remove
            Return _rows.Remove(item)
        End Function

        Public Sub RemoveAt(ByVal index As Integer) Implements System.Collections.Generic.IList(Of LookUpRow).RemoveAt
            _rows.RemoveAt(index)
        End Sub

#End Region

#Region "Helper"

        Public Function GetList() As String

            Dim buffer As New StringBuilder
            If _rows.Count < 1 Then Return ""
            For Each row As LookUpRow In _rows
                buffer.Append(Trim(row.value))
                buffer.Append(";")
            Next
            If (buffer.Length > 0) Then
                buffer = buffer.Remove(buffer.Length - 1, 1)
            End If

            Return buffer.ToString()

        End Function

        Public Sub SetList(ByVal strVals As String)

            _rows = New List(Of LookUpRow)
            If (strVals.Length <> 0) Then
                Dim strItems() As String = strVals.Split(";")
                Dim i As Integer
                For i = 0 To strItems.Length - 1
                    Dim row As New LookUpRow()
                    row.value = strItems(i)
                    row.text = ""
                    _rows.Add(row)

                Next

            End If

        End Sub

        Public Sub SetList(ByVal lst As ListBox)

            _rows = New List(Of LookUpRow)
            Dim i As Integer
            For i = 0 To lst.Items.Count() - 1

                Dim row As New LookUpRow()
                row.value = lst.Items(i).Value
                row.text = lst.Items(i).Text
                _rows.Add(row)

            Next

        End Sub

#End Region

    End Class

End Namespace

File: ALODWebUtility.CS/LookUps/LookUpTable.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Text;
using System.Web.UI.WebControls;
using ALOD.Data;

namespace ALODWebUtility.LookUps
{
    public class LookUpTable : IList<LookUpRow>
    {
        protected SqlDataStore _adapter;
        protected List<LookUpRow> _rows;
        protected System.Web.Caching.Cache cacheSource;

        public LookUpTable()
        {
            _rows = new List<LookUpRow>();
        }

        public List<LookUpRow> rows
        {
            get
            {
                return _rows;
            }
            set
            {
                _rows = value;
            }
        }

        protected System.Web.Caching.Cache Cache
        {
            get
            {
                return System.Web.HttpContext.Current.Cache;
            }
        }

        protected SqlDataStore DataStore
        {
            get
            {
                if (_adapter == null)
                {
                    _adapter = new SqlDataStore();
                }
                return _adapter;
            }
        }

        public void AddRow(string value, string text)
        {
            LookUpRow row = new LookUpRow();
            row.value = value;
            row.text = text;
            _rows.Add(row);
        }

        public LookUpRow Find(string value)
        {
            foreach (LookUpRow row in _rows)
            {
                if (row.value == value)
                {
                    return row;
                }
            }

            return null;
        }

        public void LookUpRowReader(SqlDataStore adapter, IDataReader reader)
        {
            LookUpRow row = new LookUpRow();
            if (!reader.IsDBNull(0))
            {
                row.value = reader[0].ToString();
            }
            else
            {
                row.value = string.Empty;
            }

            if (!reader.IsDBNull(1))
            {
                row.text = reader[1].ToString();
            }
            else
            {
                row.text = string.Empty;
            }
            _rows.Add(row);
        }

        #region IList

        public int Count
        {
            get
            {
                return _rows.Count;
            }
        }

        public bool IsReadOnly
        {
            get
            {
                return false;
            }
        }

        public LookUpRow this[int index]
        {
            get
            {
                return _rows[index];
            }
            set
            {
                _rows[index] = value;
            }
        }

        public void Add(LookUpRow item)
        {
            _rows.Add(item);
        }

        public void Clear()
        {
            _rows.Clear();
        }

        public bool Contains(LookUpRow item)
        {
            return _rows.Contains(item);
        }

        public void CopyTo(LookUpRow[] array, int arrayIndex)
        {
            _rows.CopyTo(array, arrayIndex);
        }

        public IEnumerator<LookUpRow> GetEnumerator()
        {
            return _rows.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _rows.GetEnumerator();
        }

        public int IndexOf(LookUpRow item)
        {
            return _rows.IndexOf(item);
        }

        public void Insert(int index, LookUpRow item)
        {
            _rows.Insert(index, item);
        }

        public bool Remove(LookUpRow item)
        {
            return _rows.Remove(item);
        }

        public void RemoveAt(int index)
        {
            _rows.RemoveAt(index);
        }

        #endregion

        #region Helper

        public string GetList()
        {
            StringBuilder buffer = new StringBuilder();
            if (_rows.Count < 1) return "";
            foreach (LookUpRow row in _rows)
            {
                buffer.Append(row.value.Trim());
                buffer.Append(";");
            }
            if (buffer.Length > 0)
            {
                buffer = buffer.Remove(buffer.Length - 1, 1);
            }

            return buffer.ToString();
        }

        public void SetList(string strVals)
        {
            _rows = new List<LookUpRow>();
            if (strVals.Length != 0)
            {
                string[] strItems = strVals.Split(';');
                int i;
                for (i = 0; i <= strItems.Length - 1; i++)
                {
                    LookUpRow row = new LookUpRow();
                    row.value = strItems[i];
                    row.text = "";
                    _rows.Add(row);
                }
            }
        }

        public void SetList(ListBox lst)
        {
            _rows = new List<LookUpRow>();
            int i;
            for (i = 0; i <= lst.Items.Count - 1; i++)
            {
                LookUpRow row = new LookUpRow();
                row.value = lst.Items[i].Value;
                row.text = lst.Items[i].Text;
                _rows.Add(row);
            }
        }

        #endregion
    }
}
